Posts/2022/07/Password Expiry Notification Using Teams and Graph API

```diff
@@ -0,0 +1,277 @@
+
+```powershell-console
+PS> Get-MgContext
+
+ClientId              : 2ee82eec-204b-204b-204b-e55eec26bf5a
+TenantId              : 14d82eec-4d5a-4d5a-4d5a-26bf5ae55eec
+CertificateThumbprint : 
+Scopes                : {User.Read.All, User.ReadWrite.All…}
+AuthType              : Delegated
+AuthProviderType      : InteractiveAuthenticationProvider
+CertificateName       : 
+Account               : farismalaeb@contoso.com
+AppName               : Microsoft Graph PowerShell
+ContextScope          : CurrentUser
+Certificate           : 
+PSHostVersion         : 2022.6.3
+ClientTimeout         : 00:05:00
+```
+
+To add additional permission to the scope, rerun the `Connect-MgGraph`, set the new scope to the **Scope** parameter and connect again. There is no need to specify the same scope already provided.
+
+So let's assume that an admin connected to the Graph API using the following scope
+
+```powershell
+$Scope=@('User.Read.All')
+Connect-MgGraph -Scopes $Scope
+```
+
+Later on, the admin wants to add the **User.ReadWrite.All** permission, all the admin needs to do is run the `Connect-MgGraph` and set the new scope
+
+```powershell
+$Scope=@('User.ReadWrite.All')
+Connect-MgGraph -Scopes $Scope
+```
+
+The new scope permission adds to the current one.
+
+A good starting point in finding out the required permission to execute a certain cmdlet is [Find-MgGraphcommand](https://docs.microsoft.com/en-us/powershell/microsoftgraph/find-mg-graph-command?view=graph-powershell-1.0) and [Graph Explorer](https://developer.microsoft.com/en-us/graph/graph-explorer).
+
+## **Password Expiry Notification Using Teams and Graph API**
+
+### Getting Password Expiration information
+
+To make this tutorial more fun, let's make a scenario.
+Consider a company named Contoso.com.
+Contoso.com have an on-premise AD syncing with AAD. The Password Expiration policy is set to 3 months.
+
+The administrator is looking for a way to send the users a notification through Microsoft Teams chat one week before the password expires, so how to start.
+
+To get the password expiration for users, use the following code.
+This code reads the **Name**, **EmailAddress**, **UserPrincipalName** and **[msDS-UserPasswordExpiryTimeComputed](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/f9e9b7e2-c7ac-4db6-ba38-71d9696981e9)**.
+The **msDS-UserPasswordExpiryTimeComputed** return when the users password expires, check it below.
+
+```powershell
+$DaysToSendWarning = (Get-Date).AddDays(7).ToLongDateString()
+**$Users = Get-ADUser -Filter {Enabled -eq $True -and PasswordNeverExpires -eq $False -and PasswordLastSet -gt 0 } `
+    -Properties "Name", "EmailAddress", "msDS-UserPasswordExpiryTimeComputed","UserPrincipalName" -SearchBase $LDAPdistinguishedName | Select-Object -Property "Name", "UserPrincipalName","EmailAddress", `
+    @{Name = "PasswordExpiry"; Expression = {[datetime]::FromFileTime($_."msDS-UserPasswordExpiryTimeComputed").ToLongDateString() }}**
```

Comment:
Reworked things a little to use splatting/limit line lengths as much as possible for readers.

```suggestion
$DaysToSendWarning = (Get-Date).AddDays(7).ToLongDateString()

$QueryParameters = @{
    Filter     = {
        Enabled -eq $true -and
        PasswordNeverExpires -eq $false -and
        PasswordLastSet -gt 0
    }
    Properties = @(
        'Name'
        'EmailAddress'
        'msDS-UserPasswordExpiryTimeComputed'
        'UserPrincipalName'
    )
    SearchBase = $LDAPdistinguishedName
}

$SelectionProperties = @(
    "Name"
    "UserPrincipalName"
    "EmailAddress"
    @{
        Name = 'PasswordExpiry'
        Expression = {
            [datetime]::FromFileTime($_."msDS-UserPasswordExpiryTimeComputed").ToLongDateString()
        }
    }
)

$Users = Get-ADUser @QueryParameters | Select-Object -Property $SelectionProperties
```